### PR TITLE
Add nullable types

### PR DESCRIPTION
**Pull request by Furgas**
`T::nullable()` wrapper:
```php
$point = new Tuple(T::float(), T::float(), T::nullable(T::generic(Post::class)));

$point[0] = 1.5;
$point[1] = 3;
$point[2] = null;
```

### CODE REVIEW

**Comment by brendt** [comment]
Hi @Furgas! Looks interesting! One question though, what's up with all these changes?

```
-            if (! isset($data[$name])) {
+            if (! array_key_exists($name, $data)) {
```

**Comment by brendt** [comment]
I'd also like your input on different syntaxes. Your approach is definitely the cleanest one, so we'll merge that one in.

Before doing that, maybe it would be good to think about some "shorthands"? I don't know how I feel about these, what's your opinion?

```php
T::nullable(T::generic(Post::class))

// vs

T::generic(Post::class)->nullable()

// vs

T::generic(Post::class, null)
T::int(null)

// vs

nullable(T::float())
```

**Comment by Furgas** [comment]
I don't have strong preference (maybe not the last one as I'm not coming from Laravel). Also I would rather make `T::int(bool $is_nullable = false)`. 
But thinking about it, this feature could be also done differently (obviously). That is by introducing `Type::isNullable(): bool` method and a flag in constructors and `T` methods.

```php
final class IntegerType implements Type
{
    /**
     * @var bool
     */
    private $is_nullable;
    
    public function __construct(bool $is_nullable = false)
    {
        $this->is_nullable = $is_nullable;
    }

    public function nullable(): self {
        return new self(true);
    }
    
    public function __invoke(int $value): int
    {
        return $value;
    }
}
```

```php
class T
{
    public static function int(bool $is_nullable = false): IntegerType
    {
        return new IntegerType($is_nullable);
    }
}
```

But I'm not sure which one would be better.

File: src/T.php [hunk 1]
```diff
@@ -69,4 +71,9 @@ public static function string(): StringType
     {
         return new StringType();
     }
+
+    public static function nullable(Type $type): NullableType
```
Comment (Furgas) [inline src/T.php]: I'm not sure if `$type` should be typehinted. There is some inconsistency - `Collection` constructor doesn't typehint `$type` (as opposed to `Tuple`) and also `ValidatesType::resolveType` can work just with any callable (not only `Type`).

Comment (brendt) [reply src/T.php]: You can keep the typehint here, I'm still thinking about whether one or the other would be better. I liked the idea of using a closure as a type on the fly, but it makes things a little more complicated. I'll probably change everything to always be a `Type`.

**Comment by brendt** [comment]
I think `T::nullable` must stay, as it is the most expressive. Thinking about it, I don't think the `bool $is_nullable = false` approach adds clarity. Maybe `->nullable()` on every type could be like this?

```php
    public function nullable(): NullableType {
        return T::nullable($this);
    }
```

**Comment by brendt** [comment]
One more remark: maybe we could use a trait here `IsNullable` to add the functionality to types.

**Comment by Furgas** [comment]
Yeah, why not. But I'd rather do:

```php
public function nullable(): NullableType {
        return new NullableType($this);
}
```

There's no need here to introduce coupling with `T`.

Should this method go into the `Type` interface? And maybe a trait with just this method?

**Comment by brendt** [comment]
Sound good! Go for it! I'll merge it this afternoon :)

**Comment by Furgas** [comment]
I'm having hard time to understand the sort order of `use` statements that StyleCI forces...

File: src/Types/NullableType.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    public function __construct(Type $type)
+    {
+        $this->type = $type;
+    }
+
+    public function __invoke($value)
+    {
+        if ($value === null) {
+            return null;
```
Comment (Furgas) [inline src/Types/NullableType.php]: It feels weird to omit `null` here, as per StyleCI complaint.

**Comment by brendt** [comment]
Super nice of you to take care of this! Don't worry about StyleCI, I'll merge it on master :)

Thanks!

**Pull request merged**